File: sdks/csharp/PgChangeFeed.Client/Http/Models/Changes.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PgChangeFeed.Client.Http.Models;

/// <summary>
/// One persisted change as returned by <c>GET /changes</c> (SPEC-022) — thirteen
/// fields: the ten the live surfaces (gRPC, SSE, NATS) carry, plus
/// <c>commit_position</c>, <c>committed_at</c> and <c>origin</c>.
/// <see cref="OldImage"/>/<see cref="NewImage"/> carry the row image as an
/// embedded JSON value or <c>null</c> when absent; kept as
/// <see cref="JsonElement"/> rather than a fixed shape, because the row
/// image's own shape depends on the captured table, not on this wire contract.
/// <see cref="Origin"/> is <c>wal</c> for a change captured from the
/// replication stream and <c>backfill</c> for an existing-rows change
/// (LH-FA-CAP-009); it is carried as the server's string (an empty or
/// unknown value included), and a response without the field or with a JSON
/// <c>null</c> reads as <c>wal</c>. The live surfaces carry no <c>origin</c>.
/// </summary>
public sealed record Change(
    [property: JsonPropertyName("commit_position")] long CommitPosition,
    [property: JsonPropertyName("change_id")] string ChangeId,
    [property: JsonPropertyName("transaction_id")] string TransactionId,
    [property: JsonPropertyName("source_table_id")] string SourceTableId,
    [property: JsonPropertyName("schema")] string Schema,
    [property: JsonPropertyName("table")] string Table,
    [property: JsonPropertyName("sequence")] long Sequence,
    [property: JsonPropertyName("operation")] string Operation,
    [property: JsonPropertyName("old_image")] JsonElement? OldImage,
    [property: JsonPropertyName("new_image")] JsonElement? NewImage,
    [property: JsonPropertyName("schema_version")] string SchemaVersion,
    [property: JsonPropertyName("committed_at")] string CommittedAt,
    [property: JsonPropertyName("origin"), JsonConverter(typeof(OriginConverter))] string Origin = "wal");

/// <summary>
/// Reads the <c>origin</c> string as the server sent it and a JSON
/// <c>null</c> as <c>wal</c>; <c>System.Text.Json</c> skips a converter for
/// <c>null</c> unless <see cref="HandleNull"/> is set.
/// </summary>
internal sealed class OriginConverter : JsonConverter<string>
{
    public override bool HandleNull => true;

    public override string Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
        reader.TokenType == JsonTokenType.Null ? "wal" : reader.GetString()!;

    public override void Write(Utf8JsonWriter writer, string value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value);
}

/// <summary>
/// <c>ReadChanges</c> response (<c>200</c>) — an empty list on no match,
/// never <c>404</c> (SPEC-022).
/// </summary>
public sealed record ReadChangesResponse(
    [property: JsonPropertyName("changes")] IReadOnlyList<Change> Changes);

File: sdks/kotlin/pgchangefeed-kotlin/src/main/kotlin/io/github/pt9912/pgchangefeed/http/model/Changes.kt
package io.github.pt9912.pgchangefeed.http.model

import com.google.gson.JsonElement
import com.google.gson.annotations.SerializedName

/**
 * One persisted change as returned by `GET /changes` (SPEC-022) — thirteen
 * fields: the ten of the Go domain type `model.Change`, plus
 * `commit_position`, `committed_at` and `origin`. `oldImage`/`newImage`
 * carry the row image as an embedded JSON value or a JSON `null` when
 * absent (`INSERT` has no `old_image`, `DELETE` has no `new_image`); kept
 * as [JsonElement] rather than a fixed shape, because the row image's own
 * shape depends on the captured table, not on this wire contract (same
 * choice as the C#/Python siblings, `System.Text.Json.JsonElement`/`Any`).
 *
 * [origin] is `wal` for a change captured from the replication stream and
 * `backfill` for an existing-rows change (LH-FA-CAP-009); it is the
 * server's string, and a response without the field (or with a JSON
 * `null`) reads as `wal`. Gson skips Kotlin default parameters, so the
 * wire value sits in the nullable [wireOrigin] and [origin] applies the
 * default. The live surfaces (gRPC, SSE, NATS) carry no `origin`.
 *
 * Gson's `JsonElement` adapter represents a present-but-`null` JSON value
 * as `JsonElement.JsonNull` (`isJsonNull == true`), not as a Kotlin `null`
 * reference — the field type stays nullable only to cover a value entirely
 * absent from the response body (outside the documented SPEC-022 shape, in
 * which the key is always present). A caller checks "no row image" via
 * `oldImage?.isJsonNull != false`, not via a plain null check.
 */
data class Change(
    @SerializedName("commit_position") val commitPosition: Long,
    @SerializedName("change_id") val changeId: String,
    @SerializedName("transaction_id") val transactionId: String,
    @SerializedName("source_table_id") val sourceTableId: String,
    @SerializedName("schema") val schema: String,
    @SerializedName("table") val table: String,
    @SerializedName("sequence") val sequence: Long,
    @SerializedName("operation") val operation: String,
    @SerializedName("old_image") val oldImage: JsonElement?,
    @SerializedName("new_image") val newImage: JsonElement?,
    @SerializedName("schema_version") val schemaVersion: String,
    @SerializedName("committed_at") val committedAt: String,
    @SerializedName("origin") val wireOrigin: String? = null,
) {
    val origin: String
        get() = wireOrigin ?: "wal"
}

/** `ReadChanges` response (`200`) — an empty list on no match, never `404` (SPEC-022). */
data class ReadChangesResponse(
    @SerializedName("changes") val changes: List<Change>,
)
